sammba/registration/struct.py

```diff
@@ -387,130 +418,85 @@ def anats_to_common(anat_filenames, write_dir, brain_volume,
-                suffixed_file = fname_presuffix(
-                    centered_head_file,
-                    suffix='_warped{}'.format(n_iter + 3))
-                if n_iter == len(nonlinear_levels):
-                    out_file = os.path.join(write_dir,
-                                            os.path.basename(suffixed_file))
-                else:
-                    out_file = suffixed_file
+        out_file = fname_presuffix(centered_head_file, suffix='_warped{}'.format(n_iter))
+    
+        for warp_file, centered_head_file in zip(previous_warp_files, 
+                                                 centered_head_files):
 
+            if n_iter == 0:
+                out_qwarp = qwarp(
+                    in_file=centered_head_file,
+                    base_file=out_tstat_allineated_head.outputs.out_file,
+                    noneg=True,
+                    iwarp=True,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
+                    inilev=0,
+                    maxlev=levels_minpatches[n_iter],
+                    out_file=out_file)
+                
+            if n_iter != 0 and n_iter < len(nonlinear_levels):
                 out_qwarp = qwarp(
                     in_file=centered_head_file,
                     base_file=out_tstat_warp_head.outputs.out_file,
-                    nmi=True,
                     noneg=True,
                     iwarp=True,
-                    weight=out_mask_tool.outputs.out_file,
-                    iniwarp=[warp_file],
-                    inilev=inilev,
-                    minpatch=nonlinear_minimal_patch,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
+                    inilev=levels_minpatches[n_iter - 1] + 1,
+                    maxlev=levels_minpatches[n_iter],
                     out_file=out_file)
-                warped_files.append(out_qwarp.outputs.warped_source)
-                warp_files.append(out_qwarp.outputs.source_warp)
-
-            out_tcat = tcat(
-                in_files=warped_files,
-                out_file=os.path.join(
-                    write_dir,
-                    'warped_{0}iters_hetemplate_filenameads.nii.gz'.format(
-                        n_iter + 3)))
-            out_tstat_warp_head = tstat(in_file=out_tcat.outputs.out_file,
-                                        outputtype='NIFTI_GZ')
+                    
+            if n_iter != 0 and n_iter >= len(nonlinear_levels):
+                out_qwarp = qwarp(
+                    in_file=centered_head_file,
+                    base_file=out_tstat_warp_head.outputs.out_file,
+                    noneg=True,
+                    iwarp=True,
+                    weight=nonlinear_weight,
+                    iniwarp=warp_file,
+                    inilev=nonlinear_levels[-1] + 1, # not ideal
+                    minpatch=levels_minpatches[n_iter],
```

Comment:
same here, you can use `level_or_minpatch` instead of `levels_minpatches[n_iter]`

Reply:
so if a user enters as `nonlinear_minpatches = [75]` and leaves the `nonlinear_levels` to default, we the two last steps will be
qwarp3: inilev=3, maxlev=3, iniwarp=warp from qwarp2
qwarp4: inilev=4, minpatch=75, inilev=warp from qwarp3

So again, why not to perform from inilev 3 to minpatch 75 ?

Reply:
it is safer to defaulting to the use of levels as much as possible since we have not fully determined the image size / level / patch size relationship

Reply:
Ok